aten/src/ATen/native/ReduceOpsUtils.h

```diff
@@ -148,11 +148,34 @@ static void allocate_reduction_result(
       }
     }
   }
-  if (result.defined()) {
-    result.resize_(shape);
-  } else {
-    result = at::empty(shape, self.options().dtype(dtype));
+  TORCH_CHECK(result.defined(), "Cannot create a new tensor inside a reduction op. You likely tried to call an operator with an out argument but the out argument was an undefined tensor.");
+  result.resize_(shape);
+}
+
+inline Tensor create_reduction_result(
+  const Tensor& self, IntArrayRef dim, bool keepdim, const c10::optional<ScalarType>& opt_dtype
+) {
+  DimMask mask = make_dim_mask(dim, self.dim());
+
+  ScalarType dtype = value_or_else(opt_dtype, [&] {
+    ScalarType src_type = self.scalar_type();
+    if (at::isIntegralType(src_type, /*includeBool=*/true)) {
+      return kLong;
+    }
+    return src_type;
+  });
+
+  auto shape = DimVector(self.sizes());
```

Comment:
Probably best to comment that this is a straight copy paste of the block above, or better yet, factor it into its own function.

Reply:
+1 to factoring